<a href="https://colab.research.google.com/github/BbekShr/Coding_exercise/blob/main/3.%20Data%20quality%20evaluation/Data_Evaluation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Data preparation of the following datasets:
1. Users.json
2. Brands.json
3. Receipts.json


---


**First step:** Import and preview all data

---


**Second step:** Clean and normalize data

---

**Third step:** Create a new dataset


*   from Receipts.json > rewardsReceiptItemList so that the list is easier to manage
*   from Brands.json > cpg so that list are easier to manager if it grows

---
**Fourth step:** Export cleaned data to csv to run queries



---

**Fifth step: Explore data qualities issue** ( Usually, I would recommend doing this before answering business question and run queries, but at the same time I would like to bring the issues to your attention regaring my findings)



In [20]:
#Import datasets

from google.colab import drive
import pandas as pd

#mount google drive
drive.mount('/content/drive')

# Define file paths
file_paths = {
    "brands": "/content/drive/MyDrive/Colab Notebooks/Data folder/brands.json",
    "receipts": "/content/drive/MyDrive/Colab Notebooks/Data folder/receipts.json",
    "users": "/content/drive/MyDrive/Colab Notebooks/Data folder/users.json",
}

# Create a dictionary to store the DataFrames
dataframes = {}

# Load each file into a DataFrame and store it in the dictionary
for name, path in file_paths.items():
    try:
        # Load the JSON file into a DataFrame
        dataframes[name] = pd.read_json(path, lines=True)
        print(f"{name} DataFrame loaded successfully.")
    except FileNotFoundError:
        print(f"File not found: {path}")
    except Exception as e:
        print(f"An error occurred while loading {name}: {e}")


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
brands DataFrame loaded successfully.
receipts DataFrame loaded successfully.
users DataFrame loaded successfully.


In [21]:
# Brand dataset
brands_df = dataframes.get('brands')
brands_df.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name  topBrand                      brandCode  
0  test brand @1612366101024       0.0                            NaN  
1                  Starbucks       0.0                      STARBUCKS  
2  test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176  
3  test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051  
4  test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827

In [22]:
# receipts dataset
receipts_df = dataframes.get('receipts')
receipts_df.head()

_id  bonusPointsEarned  \
0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4  {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate  pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                       NaN           5.0  {'$date': 1609632000000}   
3  {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6  
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052

In [23]:
# users dataset
users_df = dataframes.get('users')
users_df.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

**Clean and normalize data**

In [24]:
#Users.json file


def convert_timestamp_to_date(df, column_names):
    for column_name in column_names:
        df[column_name] = pd.to_datetime(
            df[column_name].apply(lambda x: x['$date'] if isinstance(x, dict) else x), unit='ms').dt.round('s')
    return df

def unnest_columns(df, column_names):
    for column_name in column_names:
        df[column_name] = df[column_name].apply(lambda x: x['$oid'] if isinstance(x, dict) and '$oid' in x else x)
    return df

columns_to_convert = ['createdDate', 'lastLogin']
users_df = convert_timestamp_to_date(users_df, columns_to_convert)

columns_to_unnest = ['_id']
users_df = unnest_columns(users_df, columns_to_unnest)

# Preview the DataFrame
users_df.head()

_id  active         createdDate           lastLogin  \
0  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:05 2021-01-03 15:25:38   
1  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:05 2021-01-03 15:25:38   
2  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:05 2021-01-03 15:25:38   
3  5ff1e1eacfcf6c399c274ae6    True 2021-01-03 15:25:31 2021-01-03 15:25:31   
4  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:05 2021-01-03 15:25:38   

       role signUpSource state  
0  consumer        Email    WI  
1  consumer        Email    WI  
2  consumer        Email    WI  
3  consumer        Email    WI  
4  consumer        Email    WI

In [25]:
#brands.json file
columns_to_unnest = ['_id']
brands_df = unnest_columns(brands_df, columns_to_unnest)
brands_df.head()

# Create a new DataFrame with the 'cpg' column from brands_df
cpg_df = brands_df[['cpg']].copy()

# Remove the 'cpg' column from brands_df
brands_df = brands_df.drop(columns=['cpg'])

brands_df.head()

_id       barcode        category      categoryCode  \
0  601ac115be37ce2ead437551  511111019862          Baking            BAKING   
1  601c5460be37ce2ead43755f  511111519928       Beverages         BEVERAGES   
2  601ac142be37ce2ead43755d  511111819905          Baking            BAKING   
3  601ac142be37ce2ead43755a  511111519874          Baking            BAKING   
4  601ac142be37ce2ead43755e  511111319917  Candy & Sweets  CANDY_AND_SWEETS   

                        name  topBrand                      brandCode  
0  test brand @1612366101024       0.0                            NaN  
1                  Starbucks       0.0                      STARBUCKS  
2  test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176  
3  test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051  
4  test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827

**Create cpg.csv**

In [26]:
#Create cpg.csv from brands.json

# Normalize the 'cpg' column
cpg_normalized = pd.json_normalize(cpg_df['cpg'])

# Rename columns
cpg_normalized = cpg_normalized.rename(columns={'$id.$oid': '_id', '$ref': 'ref'})

# Drop the original 'cpg' column and add normalized data
cpg_df = pd.concat([cpg_df.drop(columns=['cpg']), cpg_normalized], axis=1)

# Display the updated DataFrame
cpg_df.head()

ref                       _id
0  Cogs  601ac114be37ce2ead437550
1  Cogs  5332f5fbe4b03c9a25efd0ba
2  Cogs  601ac142be37ce2ead437559
3  Cogs  601ac142be37ce2ead437559
4  Cogs  5332fa12e4b03c9a25efd1e7

In [27]:
#receipts.json file

#unnest column
columns_to_unnest = ['_id']
receipts_df = unnest_columns(receipts_df, columns_to_unnest)

columns_to_convert = ['createDate', 'dateScanned', 'finishedDate','modifyDate','pointsAwardedDate','purchaseDate' ]
receipts_df = convert_timestamp_to_date(receipts_df, columns_to_convert)

# Create a new DataFrame with the 'cpg' column from brands_df
rewardsReceiptItemList_df = receipts_df[['_id','rewardsReceiptItemList']].copy()

# Remove the 'cpg' column from brands_df
receipts_df = receipts_df.drop(columns=['rewardsReceiptItemList'])

# Display the first few rows of both DataFrames (optional)
receipts_df.head()

_id  bonusPointsEarned  \
0  5ff1e1eb0a720f0523000575              500.0   
1  5ff1e1bb0a720f052300056b              150.0   
2  5ff1e1f10a720f052300057a                5.0   
3  5ff1e1ee0a7214ada100056f                5.0   
4  5ff1e1d20a7214ada1000561                5.0   

                             bonusPointsEarnedReason          createDate  \
0  Receipt number 2 completed, bonus point schedu... 2021-01-03 15:25:31   
1  Receipt number 5 completed, bonus point schedu... 2021-01-03 15:24:43   
2                         All-receipts receipt bonus 2021-01-03 15:25:37   
3                         All-receipts receipt bonus 2021-01-03 15:25:34   
4                         All-receipts receipt bonus 2021-01-03 15:25:06   

          dateScanned        finishedDate          modifyDate  \
0 2021-01-03 15:25:31 2021-01-03 15:25:31 2021-01-03 15:25:36   
1 2021-01-03 15:24:43 2021-01-03 15:24:43 2021-01-03 15:24:48   
2 2021-01-03 15:25:37                 NaT 2021-01-03 15:25:42   
3 2021-01-03 15:25:34 2021-01-03 15:25:34 2021-01-03 15:25:39   
4 2021-01-03 15:25:06 2021-01-03 15:25:11 2021-01-03 15:25:11   

    pointsAwardedDate  pointsEarned        purchaseDate  purchasedItemCount  \
0 2021-01-03 15:25:31         500.0 2021-01-03 00:00:00                 5.0   
1 2021-01-03 15:24:43         150.0 2021-01-02 15:24:43                 2.0   
2                 NaT           5.0 2021-01-03 00:00:00                 1.0   
3 2021-01-03 15:25:34           5.0 2021-01-03 00:00:00                 4.0   
4 2021-01-03 15:25:06           5.0 2021-01-02 15:25:06                 2.0   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6  
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052

In [28]:
rewardsReceiptItemList_df.head()


_id                             rewardsReceiptItemList
0  5ff1e1eb0a720f0523000575  [{'barcode': '4011', 'description': 'ITEM NOT ...
1  5ff1e1bb0a720f052300056b  [{'barcode': '4011', 'description': 'ITEM NOT ...
2  5ff1e1f10a720f052300057a  [{'needsFetchReview': False, 'partnerItemId': ...
3  5ff1e1ee0a7214ada100056f  [{'barcode': '4011', 'description': 'ITEM NOT ...
4  5ff1e1d20a7214ada1000561  [{'barcode': '4011', 'description': 'ITEM NOT ...

**Create rewardsReceiptItemList.csv**

In [29]:
import pandas as pd
import ast

# Function to safely parse strings into lists of dictionaries
def parse_rewards(value):
    if isinstance(value, str):
        try:
            parsed_value = ast.literal_eval(value)
            # Ensure the output is always a list
            return parsed_value if isinstance(parsed_value, list) else [parsed_value]
        except (ValueError, SyntaxError):
            return []  # Return an empty list if parsing fails
    elif isinstance(value, list):
        return value  # Return as-is if it's already a list
    else:
        return []

# Apply the parsing function to the 'rewardsReceiptItemList' column
rewardsReceiptItemList_df['rewardsReceiptItemList'] = rewardsReceiptItemList_df['rewardsReceiptItemList'].apply(parse_rewards)

expanded_data = []
for idx, row in rewardsReceiptItemList_df.iterrows():
    for item in row['rewardsReceiptItemList']:
        item['_id'] = row['_id']  # Add the '_id' from the original row to each item
        expanded_data.append(item)
normalized_rewards = pd.json_normalize(expanded_data)

# Display the normalized DataFrame with '_id'
normalized_rewards.head()

barcode                                        description finalPrice  \
0          4011                                     ITEM NOT FOUND      26.00   
1          4011                                     ITEM NOT FOUND          1   
2  028400642255  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...      10.00   
3           NaN                                                NaN        NaN   
4          4011                                     ITEM NOT FOUND      28.00   

  itemPrice needsFetchReview partnerItemId preventTargetGapPoints  \
0     26.00            False             1                   True   
1         1              NaN             1                    NaN   
2     10.00             True             2                   True   
3       NaN            False             1                   True   
4     28.00            False             1                   True   

   quantityPurchased userFlaggedBarcode userFlaggedNewItem  ... itemNumber  \
0                5.0               4011               True  ...        NaN   
1                1.0                NaN                NaN  ...        NaN   
2                1.0       028400642255               True  ...        NaN   
3                NaN               4011               True  ...        NaN   
4                4.0               4011               True  ...        NaN   

   originalMetaBriteQuantityPurchased pointsEarned targetPrice  \
0                                 NaN          NaN         NaN   
1                                 NaN          NaN         NaN   
2                                 NaN          NaN         NaN   
3                                 NaN          NaN         NaN   
4                                 NaN          NaN         NaN   

  competitiveProduct originalFinalPrice originalMetaBriteItemPrice deleted  \
0                NaN                NaN                        NaN     NaN   
1                NaN                NaN                        NaN     NaN   
2                NaN                NaN                        NaN     NaN   
3                NaN                NaN                        NaN     NaN   
4                NaN                NaN                        NaN     NaN   

  priceAfterCoupon metabriteCampaignId  
0              NaN                 NaN  
1              NaN                 NaN  
2              NaN                 NaN  
3              NaN                 NaN  
4              NaN                 NaN  

[5 rows x 35 columns]

**Export to CSV to USE in SQL**

In [30]:
#Export to CSV

users_df.to_csv('/content/drive/MyDrive/Colab Notebooks/cleaned_data/users.csv', index=False)
brands_df.to_csv('/content/drive/MyDrive/Colab Notebooks/cleaned_data/brands.csv', index=False)
cpg_df.to_csv('/content/drive/MyDrive/Colab Notebooks/cleaned_data/cpg.csv', index=False)
receipts_df.to_csv('/content/drive/MyDrive/Colab Notebooks/cleaned_data/receipts.csv', index=False)
normalized_rewards.to_csv('/content/drive/MyDrive/Colab Notebooks/cleaned_data/rewardsReceiptItemList.csv', index=False)


**Explore data qualities issue**


In [31]:
import pandas as pd

# Addint dataframes to a list
dataframes = {
    "users_df": users_df,
    "brands_df": brands_df,
    "cpg_df": cpg_df,
    "receipts_df": receipts_df,
    "normalized_rewards": normalized_rewards
}

# Function to perform basic data quality checks
def evaluate_data_quality(df_name, df):
    print(f"--- Evaluating {df_name} ---")
    results = {}

    # Check for missing values
    results["missing_values"] = df.isnull().sum()

    # Check for duplicate rows
    results["duplicate_rows"] = df.duplicated().sum()

    # Print results
    for key, value in results.items():
        print(f"{key}:")
        print(value)
        print("\n")

# Iterate through all DataFrames in the list
for name, df in dataframes.items():
    evaluate_data_quality(name, df)

--- Evaluating users_df ---
missing_values:
_id              0
active           0
createdDate      0
lastLogin       62
role             0
signUpSource    48
state           56
dtype: int64


duplicate_rows:
283


--- Evaluating brands_df ---
missing_values:
_id               0
barcode           0
category        155
categoryCode    650
name              0
topBrand        612
brandCode       234
dtype: int64


duplicate_rows:
0


--- Evaluating cpg_df ---
missing_values:
ref    0
_id    0
dtype: int64


duplicate_rows:
963


--- Evaluating receipts_df ---
missing_values:
_id                          0
bonusPointsEarned          575
bonusPointsEarnedReason    575
createDate                   0
dateScanned                  0
finishedDate               551
modifyDate                   0
pointsAwardedDate          582
pointsEarned               510
purchaseDate               448
purchasedItemCount         484
rewardsReceiptStatus         0
totalSpent                 435
userId             

Dataset


1.	**Users**: There are 283 duplicate values, which seems reasonable since users can log in multiple times. Missing states also appear normal, as some countries do not have states. However, there are a few missing SignUpSource entries. This might be because they belong to returning users, but it would be interesting to investigate why these values are missing, as they could provide insights into user acquisition channels.

2.	**Brands**: There are 234 missing brandCode values. Since brandCode is useful for analyzing how often specific brands appear in receipts, this could impact trend analysis. Alternatively, using barcode as a fallback may help fill the gap in understanding brand trends.

3.	**CPG**: This dataset is derived from Brands, so duplicates shouldn’t be an issue. Most entries seem to have consistent CPG values, which aligns with expectations.

4.	**Receipts**: There are no duplicate rows, which is great. However, some missing values for purchasedItemCount and purchaseDate limit the insights we can gain from the data. These values could potentially be extracted from each receipt to provide a more comprehensive view of purchasing trends.

5.	**RewardsReceiptItemList** (normalized_rewards): While it’s expected to have some missing rows, the absence of critical data such as brandCode, barcode, and item price reduces the ability to analyze user trends effectively. This information is vital for understanding which brands users are buying and how much they are spending.


In [32]:
# List of columns that can be null but not negative
columns_to_check = ["bonusPointsEarned", "purchasedItemCount", "finalPrice", "itemPrice", "targetPrice"]

# Function to check for negative values
def validate_no_negative_values(df, columns):
    results = {}
    for column in columns:
        if column in df.columns:
            # Convert to numeric and handle non-numeric values as NaN
            df[column] = pd.to_numeric(df[column], errors='coerce')

            # Count negative values, ignoring NaN values
            negative_count = (df[column] < 0).sum()
            results[column] = negative_count
    return results

# Iterate through all DataFrames in the dictionary
for name, df in dataframes.items():
    print(f"Evaluating data quality for: {name}")
    negative_value_issues = validate_no_negative_values(df, columns_to_check)
    print("Summary of negative value issues:")
    print(negative_value_issues)


Evaluating data quality for: users_df
Summary of negative value issues:
{}
Evaluating data quality for: brands_df
Summary of negative value issues:
{}
Evaluating data quality for: cpg_df
Summary of negative value issues:
{}
Evaluating data quality for: receipts_df
Summary of negative value issues:
{'bonusPointsEarned': 0, 'purchasedItemCount': 0}
Evaluating data quality for: normalized_rewards
Summary of negative value issues:
{'finalPrice': 0, 'itemPrice': 0, 'targetPrice': 0}


Check for columns such as ["bonusPointsEarned", "purchasedItemCount", "finalPrice", "itemPrice", "targetPrice"] that cannot be less than zero. All columns passed the test